engine/consensus/matching/core.go

```diff
@@ -695,6 +699,71 @@ func (c *Core) sealableResults() ([]*flow.IncorporatedResult, nextUnsealedResult
 	return results, nextUnsealeds, nil
 }
 
+// resultHasMultipleReceipts implements an additional _temporary_ safety measure:
+// only consider incorporatedResult sealable if we have AT LEAST 2 receipts committing
+// to the respective result from _different_ ENs
+func (c *Core) resultHasMultipleReceipts(incorporatedResult *flow.IncorporatedResult) bool {
+	blockID := incorporatedResult.Result.BlockID // block the result pertains to
+	resultID := incorporatedResult.Result.ID()
+
+	// get all receipts that are known for the block
+	receipts, err := c.receiptsDB.ByBlockIDAllExecutionReceipts(blockID)
+	if err != nil {
+		log.Error().Err(err).
+			Hex("block_id", logging.ID(blockID)).
+			Msg("could not get receipts by block ID")
+		return false
+	}
+
+	// index receipts for given incorporatedResult by their executor
+	// in case there are multiple receipts from the same executor, we keep the last one
+	receiptByExecutor := make(map[flow.Identifier]*flow.ExecutionReceipt) //map: executorID -> receipt
+	for _, receipt := range receipts {
+		if resultID != receipt.ExecutionResult.ID() { // skip receipts that commit to _different_ results
+			continue
+		}
+		receiptByExecutor[receipt.ExecutorID] = receipt
+	}
+
+	// to few receipts
+	if len(receiptByExecutor) < 2 {
+		return false
+	}
+	// At this point, we have at least two receipts from different executors and should be able to
+	// generate a candidate seal. We log information about all receipts for the result as evidence.
+	// Note that this code is probably run many times (until the seal has been finalized).
+	// However, as soon as we have at least 2 receipts, the matching Engine will produce a candidate
+	// seal and put it in the mempool. To prevent spamming the logs, we log only when there is
+	// no seal in the IncorporatedResultSeals mempool:
+	_, sealAlreadyGenerated := c.seals.ByID(incorporatedResult.ID()) // IncorporatedResultSeal has the same ID as incorporatedResult
```

Comment:
This caused some confusion in the [previous version of this PR](https://github.com/onflow/flow-go/pull/444) (see [comments](https://github.com/onflow/flow-go/pull/444#discussion_r580448616)). I have revised the comment in response, specifically this part: https://github.com/onflow/flow-go/blob/5df7490665b3ee353e15fc3f02ba44aa827801be/engine/consensus/matching/core.go#L736-L738

As a side comment: the naming on matching engine is a bit different: the persistent storage components have names ending in `DB`, such as: https://github.com/onflow/flow-go/blob/86124f6dbfc7138029e73751232ab6bcc27b7910/engine/consensus/matching/engine.go#L68

Components _without_ `DB` refer to mempools. Specifically, `e.seals` has the following type: 
https://github.com/onflow/flow-go/blob/86124f6dbfc7138029e73751232ab6bcc27b7910/engine/consensus/matching/engine.go#L74

It stores `IncorporatedResultSeal`s,  whose  ID method is defined as
https://github.com/onflow/flow-go/blob/86124f6dbfc7138029e73751232ab6bcc27b7910/model/flow/incorporated_result_seal.go#L16-L20